In [4]:
import pandas as pd
import numpy as np

In [86]:
data = pd.read_csv("./IPL Matches 2008-2020.csv")
ball_data = pd.read_csv("./IPL Ball-by-Ball 2008-2020.csv")

print(data.head(2))

       id        city        date player_of_match  \
0  335982   Bangalore  2008-04-18     BB McCullum   
1  335983  Chandigarh  2008-04-19      MEK Hussey   

                                        venue  neutral_venue  \
0                       M Chinnaswamy Stadium              0   
1  Punjab Cricket Association Stadium, Mohali              0   

                         team1                  team2  \
0  Royal Challengers Bangalore  Kolkata Knight Riders   
1              Kings XI Punjab    Chennai Super Kings   

                   toss_winner toss_decision                 winner result  \
0  Royal Challengers Bangalore         field  Kolkata Knight Riders   runs   
1          Chennai Super Kings           bat    Chennai Super Kings   runs   

   result_margin eliminator method    umpire1      umpire2  
0          140.0          N    NaN  Asad Rauf  RE Koertzen  
1           33.0          N    NaN  MR Benson   SL Shastri  


### Find the null values in each column

In [87]:
data.isna().sum()

id                   0
city                13
date                 0
player_of_match      4
venue                0
neutral_venue        0
team1                0
team2                0
toss_winner          0
toss_decision        0
winner               4
result               4
result_margin       17
eliminator           4
method             797
umpire1              0
umpire2              0
dtype: int64

In [8]:
ball_data.isna().sum()

id                       0
inning                   0
over                     0
ball                     0
batsman                  0
non_striker              0
bowler                   0
batsman_runs             0
extra_runs               0
total_runs               0
non_boundary             0
is_wicket                0
dismissal_kind      183973
player_dismissed    183973
fielder             186684
extras_type         183235
batting_team             0
bowling_team           191
dtype: int64

### Merge both datasets

In [9]:
merged = pd.merge(data, ball_data, on="id")

In [88]:
merged.head(2)

id       city        date player_of_match                  venue  \
0  335982  Bangalore  2008-04-18     BB McCullum  M Chinnaswamy Stadium   
1  335982  Bangalore  2008-04-18     BB McCullum  M Chinnaswamy Stadium   

   neutral_venue                        team1                  team2  \
0              0  Royal Challengers Bangalore  Kolkata Knight Riders   
1              0  Royal Challengers Bangalore  Kolkata Knight Riders   

                   toss_winner toss_decision  ... total_runs non_boundary  \
0  Royal Challengers Bangalore         field  ...          1            0   
1  Royal Challengers Bangalore         field  ...          1            0   

   is_wicket dismissal_kind player_dismissed fielder extras_type  \
0          0            NaN              NaN     NaN         NaN   
1          0            NaN              NaN     NaN         NaN   

            batting_team                 bowling_team  Season  
0  Kolkata Knight Riders  Royal Challengers Bangalore    2008  
1  Kolkata Knight Riders  Royal Challengers Bangalore    2008  

[2 rows x 35 columns]

### Find the null values in each column again

In [12]:
merged.isna().sum()

id                       0
city                  3139
date                     0
player_of_match        372
venue                    0
neutral_venue            0
team1                    0
team2                    0
toss_winner              0
toss_decision            0
winner                 372
result                 372
result_margin         3597
eliminator             372
method              190260
umpire1                  0
umpire2                  0
inning                   0
over                     0
ball                     0
batsman                  0
non_striker              0
bowler                   0
batsman_runs             0
extra_runs               0
total_runs               0
non_boundary             0
is_wicket                0
dismissal_kind      183973
player_dismissed    183973
fielder             186684
extras_type         183235
batting_team             0
bowling_team           191
dtype: int64

### Check info and shape after merging

In [13]:
merged.shape

(193468, 34)

In [14]:
merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 193468 entries, 0 to 193467
Data columns (total 34 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   id                193468 non-null  int64  
 1   city              190329 non-null  object 
 2   date              193468 non-null  object 
 3   player_of_match   193096 non-null  object 
 4   venue             193468 non-null  object 
 5   neutral_venue     193468 non-null  int64  
 6   team1             193468 non-null  object 
 7   team2             193468 non-null  object 
 8   toss_winner       193468 non-null  object 
 9   toss_decision     193468 non-null  object 
 10  winner            193096 non-null  object 
 11  result            193096 non-null  object 
 12  result_margin     189871 non-null  float64
 13  eliminator        193096 non-null  object 
 14  method            3208 non-null    object 
 15  umpire1           193468 non-null  object 
 16  umpire2           19

### Get the columns in the right datatype

In [16]:
merged.dtypes

id                    int64
city                 object
date                 object
player_of_match      object
venue                object
neutral_venue         int64
team1                object
team2                object
toss_winner          object
toss_decision        object
winner               object
result               object
result_margin       float64
eliminator           object
method               object
umpire1              object
umpire2              object
inning                int64
over                  int64
ball                  int64
batsman              object
non_striker          object
bowler               object
batsman_runs          int64
extra_runs            int64
total_runs            int64
non_boundary          int64
is_wicket             int64
dismissal_kind       object
player_dismissed     object
fielder              object
extras_type          object
batting_team         object
bowling_team         object
dtype: object

### Derive an index field and add it to the data set

In [89]:
merged.set_index("id", drop=True, append=False, inplace=False, verify_integrity=False).head(2)

city        date player_of_match                  venue  \
id                                                                     
335982  Bangalore  2008-04-18     BB McCullum  M Chinnaswamy Stadium   
335982  Bangalore  2008-04-18     BB McCullum  M Chinnaswamy Stadium   

        neutral_venue                        team1                  team2  \
id                                                                          
335982              0  Royal Challengers Bangalore  Kolkata Knight Riders   
335982              0  Royal Challengers Bangalore  Kolkata Knight Riders   

                        toss_winner toss_decision                 winner  ...  \
id                                                                        ...   
335982  Royal Challengers Bangalore         field  Kolkata Knight Riders  ...   
335982  Royal Challengers Bangalore         field  Kolkata Knight Riders  ...   

       total_runs  non_boundary is_wicket dismissal_kind player_dismissed  \
id                                                                          
335982          1             0         0            NaN              NaN   
335982          1             0         0            NaN              NaN   

       fielder  extras_type           batting_team  \
id                                                   
335982     NaN          NaN  Kolkata Knight Riders   
335982     NaN          NaN  Kolkata Knight Riders   

                       bowling_team Season  
id                                          
335982  Royal Challengers Bangalore   2008  
335982  Royal Challengers Bangalore   2008  

[2 rows x 34 columns]

### Identify the outliers in each column

In [19]:
merged.describe()

id  neutral_venue  result_margin         inning  \
count  1.934680e+05  193468.000000  189871.000000  193468.000000   
mean   7.567688e+05       0.094827      17.340895       1.482131   
std    3.060971e+05       0.292977      21.886397       0.499682   
min    3.359820e+05       0.000000       1.000000       1.000000   
25%    5.012270e+05       0.000000       6.000000       1.000000   
50%    7.292970e+05       0.000000       8.000000       1.000000   
75%    1.082628e+06       0.000000      20.000000       2.000000   
max    1.237181e+06       1.000000     146.000000       2.000000   

                over           ball   batsman_runs     extra_runs  \
count  193468.000000  193468.000000  193468.000000  193468.000000   
mean        9.177027       3.615967       1.240231       0.066414   
std         5.676848       1.807128       1.610867       0.339991   
min         0.000000       1.000000       0.000000       0.000000   
25%         4.000000       2.000000       0.000000       0.000000   
50%         9.000000       4.000000       1.000000       0.000000   
75%        14.000000       5.000000       1.000000       0.000000   
max        19.000000       9.000000       6.000000       7.000000   

          total_runs   non_boundary      is_wicket  
count  193468.000000  193468.000000  193468.000000  
mean        1.306645       0.000083       0.049078  
std         1.598802       0.009094       0.216031  
min         0.000000       0.000000       0.000000  
25%         0.000000       0.000000       0.000000  
50%         1.000000       0.000000       0.000000  
75%         1.000000       0.000000       0.000000  
max         7.000000       1.000000       1.000000

We can see that in result_margin the max is 146.0 and mean is 17.34 so there clearly must be a unusual value creating the massive difference , henceforth we can find the similar trrends in other columns too 

Here I will be only showing the result_margin.

In [47]:
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [48]:
fig = px.histogram(merged, x="result_margin")
fig.show()

### What was the count of matches played in each season

In [90]:
merged['Season'] = pd.DatetimeIndex(merged['date']).year
merged.head(1)

id       city        date player_of_match                  venue  \
0  335982  Bangalore  2008-04-18     BB McCullum  M Chinnaswamy Stadium   

   neutral_venue                        team1                  team2  \
0              0  Royal Challengers Bangalore  Kolkata Knight Riders   

                   toss_winner toss_decision  ... total_runs non_boundary  \
0  Royal Challengers Bangalore         field  ...          1            0   

   is_wicket dismissal_kind player_dismissed fielder extras_type  \
0          0            NaN              NaN     NaN         NaN   

            batting_team                 bowling_team  Season  
0  Kolkata Knight Riders  Royal Challengers Bangalore    2008  

[1 rows x 35 columns]

In [53]:
# match_per_season=data.groupby("Season")["id"].count().reset_index()
# print(match_per_season)
data['date'] = pd.to_datetime(data['date'])
data['year'] = data['date'].dt.year
data.groupby("year")["id"].count()

year
2008    58
2009    57
2010    60
2011    73
2012    74
2013    76
2014    60
2015    59
2016    60
2017    59
2018    60
2019    60
2020    60
Name: id, dtype: int64

### How many runs were scored in each season?

In [36]:
season= merged.groupby(['Season'])['total_runs'].sum()
print(season)

Season
2008    17937
2009    16320
2010    18864
2011    21154
2012    22453
2013    22541
2014    18909
2015    18332
2016    18862
2017    18769
2018    19901
2019    19400
2020    19352
Name: total_runs, dtype: int64


### Find the top 10 teams which has won the most number of matches

In [26]:
wins = data.groupby(['winner'])['winner'].count().nlargest(10)
print(wins)

winner
Mumbai Indians                 120
Chennai Super Kings            106
Kolkata Knight Riders           99
Royal Challengers Bangalore     91
Kings XI Punjab                 88
Rajasthan Royals                81
Delhi Daredevils                67
Sunrisers Hyderabad             66
Deccan Chargers                 29
Delhi Capitals                  19
Name: winner, dtype: int64


### Find the Top 10 Players With Most Player of Match

In [54]:
potm = data['player_of_match'].value_counts().nlargest(10)
print(potm)

AB de Villiers    23
CH Gayle          22
RG Sharma         18
DA Warner         17
MS Dhoni          17
SR Watson         16
YK Pathan         16
SK Raina          14
V Kohli           13
G Gambhir         13
Name: player_of_match, dtype: int64


### Find the Percentage of choosing field after winning the toss

In [83]:
percentage_to_field=(data[data["toss_decision"]=="field"]["id"].count()/data["id"].count())*100
percentage_to_field

60.78431372549019

### Find the Percentage of choosing to bat first after winning the toss

In [84]:
percentage_to_bat=(data[data["toss_decision"]=="bat"]["id"].count()/data["id"].count())*100
percentage_to_bat

39.21568627450981

### Find the ratio of Bat to Field count

In [85]:
percentage_to_bat/percentage_to_field

0.6451612903225807

### Find the count of Loss vs Wins vs Ties

In [62]:
# results = data["result"].value_counts()
# print(results)

### Who won the most tosses

In [19]:
toss_w = data["toss_winner"].value_counts().nlargest(1)
print(toss_w)

Mumbai Indians    106
Name: toss_winner, dtype: int64


### Who has hit the most number of 4's

In [78]:

four=ball_data[ball_data['batsman_runs']==4]
runs_4=four.groupby('batsman')['batsman_runs'].count().reset_index()
runs_4.columns=['Batsman','4s']
runs_4.sort_values(by='4s',ascending=False).head(1)

Batsman   4s
329  S Dhawan  591

### Who has hit the most number of 6's?

In [77]:
four=ball_data[ball_data['batsman_runs']==6]
runs_6=four.groupby('batsman')['batsman_runs'].count().reset_index()
runs_6.columns=['Batsman','6s']
runs_6.sort_values(by='6s',ascending=False).head(1)

Batsman   6s
63  CH Gayle  349